SSA/Projects/InstCombine/LLVM/EDSL.lean

```diff
@@ -106,7 +106,15 @@ def mkExpr (Γ : Ctxt (MetaLLVM φ).Ty) (opStx : MLIR.AST.Op φ) :
         | "llvm.ashr"   => pure <| Sum.inl .ashr
         | "llvm.urem"   => pure <| Sum.inl .urem
         | "llvm.srem"   => pure <| Sum.inl .srem
-        | "llvm.add"    => pure <| Sum.inl .add
+        | "llvm.add"    =>  do
+          let attr? := opStx.attrs.getAttr "overflowFlags"
+          match attr? with
+            | .none =>  pure <| Sum.inl (MOp.BinaryOp.add ⟨ false , false ⟩ )
+            | .some y => match y with
+              | (.opaque_ "llvm.overflow" "nsw") => pure <| Sum.inl (MOp.BinaryOp.add ⟨ true ,false⟩ )
+              | (.opaque_ "llvm.overflow" "nuw") => pure <| Sum.inl (MOp.BinaryOp.add ⟨ false, true ⟩ )
+              | (.opaque_ "llvm.overflow" s ) => throw <| .generic s!"The overflow flag {s} not allowed. We currently support nsw (no signed wrap) and nuw (no unsigned wrap)"
```

Comment:
This does not seem to handle the case where we make both `nsw` and `nuw`? Don't we need to have another case where we have both flags?

Reply:
You are correct. I hadn't considered that. This case actually does appear in our test data:

```mlir
  llvm.func @shl_add_lshr_flag_preservation(%arg0: i32, %arg1: i32, %arg2: i32) -> i32 {
    %0 = llvm.shl %arg0, %arg1 overflow<nuw>  : i32
    %1 = llvm.add %0, %arg2 overflow<nsw, nuw>  : i32
    %2 = llvm.lshr %1, %arg1  : i32
    llvm.return %2 : i32
  }
```